### PR TITLE
filter witnesses prior to selected 14

### PR DESCRIPTION
**Pull request by andymck**

### CODE REVIEW

File: iot_verifier/src/runner.rs [hunk 1]
```diff
@@ -504,6 +525,19 @@ fn shuffle_and_split_witnesses(
     Ok(unselected_witnesses)
 }
 
+fn filter_witness(invalid_reason: InvalidReason) -> FilterStatus {
+    match invalid_reason {
+        InvalidReason::SelfWitness => FilterStatus::Exclude,
```
Comment (mawdegroot) [inline iot_verifier/src/runner.rs:530]: Wasn't this supposed to be a Drop?

Comment (andymck) [reply iot_verifier/src/runner.rs:530]: it sure was

**Review by bbalser — APPROVED** [review]

File: iot_verifier/src/runner.rs [hunk 2]
```diff
@@ -264,19 +269,35 @@ impl Runner {
...
+                    // filter witness list
+                    // 'drop' items are dropped to the floor, never make it to s3
+                    // 'exclude' items are not permitted in last 14 but
+                    // will make it into the unselected list and thus will goto s3
+                    // 'include' items are from where the last 14 are selected
+                    // any witness with include status which doesnt make it to the last 14
+                    // will join the excluded items in the unselected list and thus will goto s3
+                    let mut excluded_witnesses  = Vec::new();
+                    let mut selected_witnesses  = Vec::new();
+                    for witness in verified_witnesses_result.verified_witnesses {
```
[the remark this answers is not in the document]
Comment (andymck) [reply iot_verifier/src/runner.rs:285]: but surely that will involve two iterations over the witness list, one for filter then one for partition ? the for loop does  all da things in one iteration

[the remark this answers is not in the document]
Comment (madninja) [reply iot_verifier/src/runner.rs:285]: clean first, fast later..

[the remark this answers is not in the document]
Comment (andymck) [reply iot_verifier/src/runner.rs]: fixed here 35480a5188b46fc8650f5475aab83611d205818d
i see the single iteration now that i have done it, prior to then i assumed would have to be done in a two steps...so yeah looks better

File: iot_verifier/src/runner.rs [hunk 3]
```diff
@@ -264,19 +269,41 @@ impl Runner {
...
+                    // any witness with include status which doesnt make it to the last 14
+                    // will join the excluded items in the unselected list and thus will goto s3
+                    let (excluded_witnesses, mut selected_witnesses) = verified_witnesses_result
                         .verified_witnesses
                         .into_iter()
-                        .filter(|witness| witness.invalid_reason != InvalidReason::SelfWitness)
-                        .collect();
-                    let unselected_witnesses = shuffle_and_split_witnesses(
+                        .filter(|witness| {
+                            matches!(filter_witness(witness.invalid_reason), FilterStatus::Drop)
```
Comment (madninja) [inline iot_verifier/src/runner.rs:288]: nice! and using matches! too.. well done

**Review by madninja — APPROVED** [review]

**Pull request merged**